File: C++/Backtracking/17 LetterCombinationsOfAPhoneNumber.cpp
// LeetCode 17. Letter Combinations of a Phone Number
// Time Complexity: O(4^n)
// Space Complexity: O(n) — recursion stack

#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;

        if (digits.empty()) {
            return result;
        }

        vector<string> mapping = {
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        string current;
        backtrack(result, current, digits, 0, mapping);
        return result;
    }

private:
    void backtrack(vector<string>& result, string& current,
                   string& digits, int index, vector<string>& mapping) {

        // Base case
        if (index == digits.length()) {
            result.push_back(current);
            return;
        }

        int digit = digits[index] - '0';

        for (char c : mapping[digit]) {
            current.push_back(c);
            backtrack(result, current, digits, index + 1, mapping);
            current.pop_back(); // backtrack
        }
    }
};

File: Java/Backtracking/17 LetterCombinationsOfAPhoneNumber.java
// LeetCode 17. Letter Combinations of a Phone Number
// Time Complexity: O(4^n)
// Space Complexity: O(n) — recursion stack

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();

        if (digits == null || digits.length() == 0) {
            return result;
        }

        String[] mapping = {
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        backtrack(result, new StringBuilder(), digits, 0, mapping);
        return result;
    }

    private void backtrack(List<String> result, StringBuilder current,
                           String digits, int index, String[] mapping) {

        // Base case: formed one complete combination
        if (index == digits.length()) {
            result.add(current.toString());
            return;
        }

        int digit = digits.charAt(index) - '0';

        for (char c : mapping[digit].toCharArray()) {
            current.append(c);
            backtrack(result, current, digits, index + 1, mapping);
            current.deleteCharAt(current.length() - 1); // backtrack
        }
    }
}
